primitives/phragmen/src/reduce.rs

```diff
@@ -0,0 +1,1065 @@
+	// because abstracting some of this code into a function/closure is super hard due to borrow
+	// checks (and most likely needs unsafe code at the end). For now I will keep it as it and
+	// refactor later.
+
+	// a flat iterator of (voter, target) over all pairs of votes. Similar to reduce_4, we loop
+	// without borrowing.
+	for assignment_index in 0..assignments.len() {
+		let voter = assignments[assignment_index].who.clone();
+
+		let mut dist_index = 0;
+		loop {
+			// A distribution could have been removed. We don't know for sure. Hence, we check.
+			let maybe_dist = assignments[assignment_index].distribution.get(dist_index);
+			if maybe_dist.is_none() {
+				// The rest of this loop is moot.
+				break;
+			}
+			let (target, _) = maybe_dist.expect("Value checked to be some").clone();
+
+			// store if they existed already.
+			let voter_id = NodeId::from(voter.clone(), NodeRole::Voter);
+			let target_id = NodeId::from(target.clone(), NodeRole::Target);
+			let voter_exists = tree.contains_key(&voter_id);
+			let target_exists = tree.contains_key(&target_id);
+
+			// create both.
+			let voter_node = tree
+				.entry(voter_id.clone())
+				.or_insert(Node::new(voter_id).into_ref())
+				.clone();
+			let target_node = tree
+				.entry(target_id.clone())
+				.or_insert(Node::new(target_id).into_ref())
+				.clone();
+
+			// If one exists but the other one doesn't, or if both does not, then set the existing
+			// one as the parent of the non-existing one and move on. Else, continue with the rest
+			// of the code.
+			match (voter_exists, target_exists) {
+				(false, false) => {
+					Node::set_parent_of(&target_node, &voter_node);
+					dist_index += 1;
+					continue;
+				}
+				(false, true) => {
+					Node::set_parent_of(&voter_node, &target_node);
+					dist_index += 1;
+					continue;
+				}
+				(true, false) => {
+					Node::set_parent_of(&target_node, &voter_node);
+					dist_index += 1;
+					continue;
+				}
+				(true, true) => { /* don't continue and execute the rest */ }
+			};
+
+			let (voter_root, voter_root_path) = Node::root(&voter_node);
+			let (target_root, target_root_path) = Node::root(&target_node);
+
+			if voter_root != target_root {
+				// swap
+				merge(voter_root_path, target_root_path);
+				dist_index += 1;
+			} else {
+				// find common and cycle.
+				let common_count = trailing_common(&voter_root_path, &target_root_path);
+
+				// because roots are the same.
+				#[cfg(feature = "std")]
+				debug_assert_eq!(
+					target_root_path.last().unwrap(),
+					voter_root_path.last().unwrap()
+				);
+				debug_assert!(common_count > 0);
+
+				// cycle part of each path will be `path[path.len() - common_count - 1 : 0]`
+				// NOTE: the order of chaining is important! it is always build from [target, ...,
+				// voter]
+				let cycle = target_root_path
+					.iter()
+					.take(target_root_path.len() - common_count + 1)
```

Comment:
```suggestion
					.rev().skip(common_count)
```

Should be the same (if the order of the final elements isn't important)

Reply:
hmm no I don't want to rev it; the order matters.

Reply:
Okay^^ 

Than I think having the calculation guarded would be nice.